### PR TITLE
Make audit logs immune to sampling

### PR DESCRIPTION
**Pull request by vrto**
### Problem

Telemetry sampling is a native observability concept that helps the emitter to reduce the volume of telemetry data to a manageable level. This is usually fine as long as we're not talking audit log - **we don't want to drop any actions that will appear in the security audit log**.

### Solution

Our [internal logging standard](https://docs.sourcegraph.com/admin/observability/logs#log-sampling) says the following:

>  The first 100 identical log entries per second will always be output, but thereafter only every 100th identical message will be output.

The key part here is the phrase **identical message**. Our [sourcegraph/log](https://github.com/sourcegraph/log) library wraps Uber's Zap, and they indeed [sample based on the log message](https://sourcegraph.com/github.com/uber-go/zap/-/blob/zapcore/sampler.go?L221). The other fields in the structured log entry are not considered. 

To make each audit log message unique, we append a ["sampling immunity token"](https://sourcegraph.sourcegraph.com/-/editor?remote_url=https%3A%2F%2Fgithub.com%2Fsourcegraph%2Fsourcegraph.git&branch=mv%2Fauditlog%2Fsampling-immunity&file=internal%2Faudit%2Faudit.go&start_row=29&start_col=20&end_row=29&end_col=20&editor=JetBrains&version=v2.0.1) in the log message.

**Note**: we can change the implementation in the future, for example, if we make `sourcegraph/log` to make a sampling decision based on the structured log fields instead of wrapping the default Zap behavior 

## Test plan

Automated integration test.

### CODE REVIEW

**Comment by vrto** [comment]
- https://github.com/orgs/sourcegraph/projects/257/views/2

File: internal/audit/integration/integration_test.go [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+	"context"
+	"strings"
+	"testing"
+
+	"github.com/sourcegraph/run"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestIntegration(t *testing.T) {
+	t.Setenv("SRC_LOG_LEVEL", "info")
```
Comment (jhchabran) [inline internal/audit/integration/integration_test.go:13]: What about adding a small comment explaining why we have to take this route to test this?

**Review by jhchabran — APPROVED** [review]

**Comment by jhchabran** [comment]
@vrto wouldn't this decision to use those token an interesting ADR to surface?

**Review by mohammadualam — APPROVED** [review]

**Review by bobheadxi — APPROVED** [review]
Clever!

File: internal/audit/audit.go [hunk 2]
```diff
@@ -25,7 +27,7 @@ func Log(ctx context.Context, logger log.Logger, record Record) {
 			log.String("X-Forwarded-For", forwardedFor(client)))))
 	fields = append(fields, record.Fields...)
 
-	logger.Info(record.Action, fields...)
+	logger.Info(fmt.Sprintf("%s (sampling immunity token: %s)", record.Action, uuid.New().String()), fields...)
```
[the remark this answers is not in the document]
Comment (bobheadxi) [reply internal/audit/audit.go:30]: a docstring here regarding why it is included in the message would be helpful too

**Comment by vrto** [comment]
@jhchabran ADR is [on my list](https://github.com/sourcegraph/security/issues/164), but I was planning to do one for the whole audit log thing.

**Pull request merged**